### PR TITLE
OTT-939: Add support for creating deal lineitems for adpod

### PR DESCRIPTION
**Pull request by Pubmatic-Dhruv-Sonone**
Changes:

1.  Adding following config parameters in settings.py
```
             # parameter to enable deal lineitem
             ENABLE_DEAL_LINEITEM  - TRUE

             # parameter to accept deal tier config and  price for each bidder.
             DEAL_CONFIG  - {"pubmatic":{"price":10,"prefix":["abc"],"mindealtier":[5]}}
  
```

1.  Create line items with {slot_no}_pwtdt targeting  for each  prefix + mindealtier combo with price as mention in deal_config
2.  LI name -  slot_no_video_Prefix_BidderName_{prefix+mindealtier}
3. LineItem Type will be SPONSORSHIP with 100 Impression goals
4.  Number of line item per slot = len(prefix) * len(mindealtier) for each bidder
5. All the LI will be created under same order.
6. partner targeting will be added if bidder name in deal_config is present in PREBID_BIDDER_CODE list

**Summary Example:** 

Screenshot 2023-09-08 at 3 40 22 PM

**GAM setup : Targeting Keys**

Screenshot 2023-09-08 at 3 51 51 PM

**GAM setup : **
Screenshot 2023-09-08 at 3 51 10 PM

### CODE REVIEW

File: tasks/add_new_openwrap_partner.py [hunk 1]
```diff
@@ -1127,6 +1136,59 @@ def main():
...
 
+  if setup_type == constant.ADPOD and deal_lineitem_enabled == True:
+        lib.print_summary(adpod_size, adpod_creative_durations, deal_config, adpod_slots, lineitem_type) 
+        ok = input('Is this correct? (y/n)\n')
+        if ok != 'y':
+            logger.info('Exiting.')
+            return
+        try:
+            # Get the user.
+          user_id = dfp.get_users.get_user_id_by_email(user_email)
```
Comment (pm-isha-bharti) [inline tasks/add_new_openwrap_partner.py:1147]: Hi @Pubmatic-Dhruv-Sonone 
There is a significant amount of duplicated code, which may require ongoing maintenance. Is there a way we can avoid this?

File: tasks/add_new_openwrap_partner.py [hunk 2]
```diff
@@ -639,6 +695,88 @@ def get_creative_file(setup_type):
...
+    if bidder_code is not None:
+        found = False
+        for bc in bidder_code:
+            if bidder == bc:
+                found = True
+                break
+        if found:
+            key_gen_obj.set_bidder_value(bidder_code, slot)
+
+    for prefix in cfg['prefix']:
```
Comment (pm-isha-bharti) [inline tasks/add_new_openwrap_partner.py:748]: if 'prefix' and 'mindealtier' strings are used at multiple places, then please add constants.

File: tasks/add_new_openwrap_partner.py [hunk 3]
```diff
@@ -639,6 +695,88 @@ def get_creative_file(setup_type):
...
+
+  line_items_config = []
+     
+  #create line item config for each price
+  for bidder, cfg in deal_config.items():
+    if bidder_code is not None:
+        found = False
+        for bc in bidder_code:
+            if bidder == bc:
+                found = True
```
Comment (pm-isha-bharti) [inline tasks/add_new_openwrap_partner.py:743]: Can we do this if `found` is not getting used anywhere
```
if bidder == bc:
    key_gen_obj.set_bidder_value(bidder_code, slot)
    break
```

File: tasks/add_new_openwrap_partner.py [hunk 4]
```diff
@@ -1127,81 +1279,88 @@ def main():
...
-      device_capabilities=device_capabilities,
-      roadblock_type=roadblock_type,
-      use_1x1=use_1x1,
-      name_start_format=color.BOLD,
-      format_end=color.END,
-      value_start_format=color.BLUE,
-    ))
 
-  if setup_type == constant.ADPOD:
+  if deal_lineitem_enabled == False:
```
Comment (pm-isha-bharti) [inline tasks/add_new_openwrap_partner.py:1283]: Can we make it non adpod for better understanding: 
`if setup_type is not constant.ADPOD:`

File: tasks/add_new_openwrap_partner.py [hunk 5]
```diff
@@ -1127,81 +1279,88 @@ def main():
...
+            lineitem_per_slot = len(prices),
+            adpod_creative_durations = adpod_creative_durations,
+            creatives_per_slot= len(adpod_creative_durations),
+            lineitem_total = len(prices)*adpod_size,
+            creatives_total= adpod_size * len(adpod_creative_durations),
+            name_start_format=color.BOLD,
+            format_end=color.END,
+            value_start_format=color.BLUE,
+        ))
+  else:
```
Comment (pm-isha-bharti) [inline tasks/add_new_openwrap_partner.py:1352]: In reference to the above comment, this should be executed for: 
`if deal_lineitem_enabled `

File: tasks/add_new_openwrap_partner.py [hunk 6]
```diff
@@ -475,7 +489,7 @@ def process_price_bucket(self, start_index, end_index, granu):
 
 def setup_partner(user_email, advertiser_name, advertiser_type, order_name, placements,
      sizes, lineitem_type, lineitem_prefix, bidder_code, prices, setup_type, creative_template, num_creatives, use_1x1,
-     currency_code, custom_targeting, same_adv_exception, device_categories, device_capabilities, roadblock_type, slot , adpod_creative_durations, creative_user_def_var):
+     currency_code, custom_targeting, same_adv_exception, device_categories, device_capabilities, roadblock_type, slot , adpod_creative_durations, creative_user_def_var, deal_config, deal_lineitem_enabled):
```
Comment (pm-isha-bharti) [inline tasks/add_new_openwrap_partner.py:492]: Default values can be set to avoid changes at multiple places: 
`deal_lineitem_enabled=None, deal_lineitem_enabled=False`

Ignore if you need to redo lot of changes.

**Comment by pm-isha-bharti** [comment]
Please update the readme for the new config introduced.

File: settings.py [hunk 7]
```diff
@@ -183,7 +183,26 @@
...
+# THis option is only for Adpod setup
+# Set ENABLE_DEAL_LINEITEM = True for creating deal line item
+# Defaults to False
+ENABLE_DEAL_LINEITEM = False
+
+# DEAL_CONFIG - configuration for creating deal lineitem
+# THis option is only for Adpod setup
+# Provide price, dealtier prefix and mindealtier values
+# Number of LineItem created = len(prefix)*len(mindealtier) for each bidder
+# LineItem wil be created with pwt_dt(dealtier)targeting with pwt_dt value as prefix[index] + mindealtier[index] for each bidder
```
Comment (pm-isha-bharti) [inline settings.py:198]: Add a comment for `DFP_LINEITEM_TYPE` to be set as `SPONSERSHIP` for Deal lineitem setup.

File: tasks/add_new_openwrap_partner.py (hunk 6 above)
Comment (Pubmatic-Dhruv-Sonone) [reply tasks/add_new_openwrap_partner.py:492]: Default False are already set at time of variable initialisation.

File: tasks/add_new_openwrap_partner.py (hunk 4 above)
Comment (Pubmatic-Dhruv-Sonone) [reply tasks/add_new_openwrap_partner.py:1313]: We are printing this summary  for Non Adpod and Adpod (non deal) line items.

File: tasks/add_new_openwrap_partner.py (hunk 5 above)
Comment (Pubmatic-Dhruv-Sonone) [reply tasks/add_new_openwrap_partner.py:1382]: Printing all those values as part of **print_summary** function for  Adpod deallineitems

File: tasks/add_new_openwrap_partner.py (hunk 2 above)
Comment (Pubmatic-Dhruv-Sonone) [reply tasks/add_new_openwrap_partner.py]: Done

File: tasks/add_new_openwrap_partner.py (hunk 3 above)
Comment (Pubmatic-Dhruv-Sonone) [reply tasks/add_new_openwrap_partner.py]: Done

File: settings.py (hunk 7 above)
Comment (Pubmatic-Dhruv-Sonone) [reply settings.py:198]: Done

File: tasks/add_new_openwrap_partner.py (hunk 4 above)
Comment (pm-isha-bharti) [reply tasks/add_new_openwrap_partner.py:1313]: I was referring to the comment on Line no 1284: 
`# Prininting non adpod lineitem summary`

**Review by pm-isha-bharti — APPROVED** [review]
LGTM

File: constant.py [hunk 8]
```diff
@@ -10,6 +10,11 @@
 JW_PLAYER="JWPLAYER"
 ADPOD="ADPOD"
 
+LI_SPONSORSHIP = "SPONSORSHIP"
```
Comment (ShriprasadM) [inline constant.py:13]: We can rename `LI_SPONSORSHIP` to `LI_TYPE`. It would be readable.

File: docs/adpod_setup.md [hunk 9]
```diff
@@ -17,6 +17,8 @@
 |`VIDEO_LENGTHS`| Adpod video creative durations.<br> Ex `VIDEO_LENGTHS = [5,10,15]` will create creatives  with durations 5, 10 and 15 secs for each slot of Adpod. | integer array|
 |`PREBID_BIDDER_CODE` | Bidder codes to target bidders with one line item. <br>Ex `PREBID_BIDDER_CODE = ['pubmatic']`. This parameter is mandatory for bidder level reporting. Set to `None` to generate line items for all partners.   | string array|
 |`OPENWRAP_BUCKET_CSV` | Set this to one of the csv file mentioned in ` Inline Header Bidding  csv` table  below. This CSV lists buckets and price granularity; it sets `pwtpb` targeting for each line item..| string |
+|`ENABLE_DEAL_LINEITEM` | Set this value to `TRUE` for creating deal line items| boolean |
+|`DEAL_CONFIG` | Configuration for creating deal lineitem. Set this if `ENABLE_DEAL_LINEITEM` is set to `TRUE`. `DEAL_CONFIG` should have price, array of prefix and array of mindealtier values. Ex: `{"pubmatic":{"price":10,"prefix":["abc"],"mindealtier":[5]}}` | object |
```
Comment (ShriprasadM) [inline docs/adpod_setup.md:21]: Also add what would be the currency here? Is it USD default?

File: settings.py [hunk 10]
```diff
@@ -174,7 +174,7 @@
 # Use this when OPENWRAP_SETUP_TYPE = "ADPOD"
 # This parameter will be used to set creative level targeting. ex s1_pwtdur = 10
 # Represents the video length parameter on UI.
-VIDEO_LENGTHS = [10,15]
+VIDEO_LENGTHS = [10]
```
Comment (ShriprasadM) [inline settings.py:177]: Why is this changed?

File: settings.py [hunk 11]
```diff
@@ -183,7 +183,28 @@
 # ex ADPOD_SLOTS = [1,2,3], will create 1st, 2nd and 3rd slot of adpod
 # ex ADPOD_SLOTS = [4,5], will create 4th and 5th slot of adpod
 # Slot numbers should be in incremental order.
-ADPOD_SLOTS = [1,2,3]
+ADPOD_SLOTS = [1]
```
Comment (ShriprasadM) [inline settings.py:186]: Why is this changed?

File: settings.py [hunk 12]
```diff
@@ -183,7 +183,28 @@
...
+# LineItem wil be created with pwtdt(dealtier)targeting with pwtdt value as prefix[index] + mindealtier[index] for each bidder
+# LineItem wil be created with pwtdt(dealtier)targeting with pwtdt value as prefix[index] + mindealtier[index] for each bidder
+# Set DFP_LINEITEM_TYPE = "SPONSORSHIP" and ENABLE_DEAL_LINEITEM = True for creating deal lineitem
+DEAL_CONFIG = {
+    "pubmatic":{
+        "price":10,
+        "prefix":["abc"],
+        "mindealtier":[5]
+    }
+}
```
Comment (ShriprasadM) [inline settings.py:207]: 1. Is this default initialization really required? Can we move it comment section?
2. can we return error and expect user is providing the deal config?

**Review by ShriprasadM — CHANGES_REQUESTED** [review]
@Pubmatic-Dhruv-Sonone : Please checkout my comments. We can connect if required

File: settings.py (hunk 12 above)
Comment (Pubmatic-Dhruv-Sonone) [reply settings.py]: Default initialisation is not required.  If required values is not provided, error is returned.
I have moved this value to comment as sample

File: settings.py (hunk 11 above)
Comment (Pubmatic-Dhruv-Sonone) [reply settings.py]: This is just a sample config. I have set it to empty now , what kind of values are expected is mentioned in the comment above

File: settings.py (hunk 10 above)
Comment (Pubmatic-Dhruv-Sonone) [reply settings.py]: This is just a sample config. I have set it to empty now , what kind of values are expected is mentioned in the comment above

File: docs/adpod_setup.md (hunk 9 above)
Comment (Pubmatic-Dhruv-Sonone) [reply docs/adpod_setup.md]: Updated the readme.

File: constant.py (hunk 8 above)
Comment (Pubmatic-Dhruv-Sonone) [reply constant.py:13]: Will rename it to LI_TYPE_SPONSORSHIP

File: tasks/add_new_openwrap_partner.py [hunk 13]
```diff
@@ -769,12 +769,16 @@ def create_deal_line_item_configs(deal_config, order_id, placement_ids, bidder_c
...
             if bidder == bc:
-                key_gen_obj.set_bidder_value(bidder_code, slot)
+                found = True
                 break
-        
-
+        if found == True:
+            key_gen_obj.set_bidder_value(bidder_code, slot)
+        else:
+            key_gen_obj.set_bidder_value(None, slot)
```
Comment (pm-isha-bharti) [inline tasks/add_new_openwrap_partner.py:780]: Why is this required?

File: tasks/add_new_openwrap_partner.py [hunk 14]
```diff
@@ -639,6 +727,90 @@ def get_creative_file(setup_type):
...
+     
+  #create line item config for each price
+  for bidder, cfg in deal_config.items():
+    if bidder_code is not None:
+        found = False
+        for bc in bidder_code:
+            if bidder == bc:
+                found = True
+                break
+        if found == True:
```
Comment (Pubmatic-Dhruv-Sonone) [inline tasks/add_new_openwrap_partner.py:777]: @pm-isha-bharti  
We will be adding the pwtpid (bidder code) targeting only if bidderName in the deal_config is also present in the bidder_code array. 

Added the `if found == True` to set the  bidder_code targeting criteria  to none in cases where pwtpid targeting is not needed.  Earlier  bidder_code once set was used for all the lineitems.
 
Added the code to address this bug:  https://inside.pubmatic.com:9443/jira/browse/OTT-1298
